File: src/tools/path_helper.zig
const std = @import("std");
const builtin = @import("builtin");
const liu = @import("liu");

var base: []const u8 = "/";
var output: std.ArrayList(u8) = undefined;
var paths: std.StringArrayHashMap(void) = undefined;

pub fn addPath(path_: []const u8) !void {
    var path = path_;
    const result = try paths.getOrPut(path);
    if (result.found_existing) return;

    if (std.mem.startsWith(u8, path, "~/")) {
        try output.appendSlice(base);
        path = path[2..];
    }

    try output.appendSlice(path);
    try output.append(':');
}

pub fn main() !void {
    var arena_ = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    var arena = arena_.allocator();

    const args = try std.process.argsAlloc(arena);
    base = args[1];

    output = std.ArrayList(u8).init(arena);
    paths = std.StringArrayHashMap(void).init(arena);

    const env_map = try std.process.getEnvMap(arena);
    const stdout = std.io.getStdOut().writer();

    const path = env_map.get("PATH") orelse return error.MissingPathVariable;
    const cfg_dir = env_map.get("CFG_DIR") orelse return error.MissingCfgVariable;

    {
        const local_path = try std.mem.concat(arena, u8, &.{ cfg_dir, "/local/path" });
        try addPath(local_path);
    }

    // linux gopath
    try addPath("/usr/local/go/bin");

    // gopath
    try addPath("~/go");
    try addPath("~/go/bin");

    try addPath("/opt/homebrew/bin");
    try addPath("/opt/homebrew/Cellar/llvm/14.0.6/bin:$PATH");

    try addPath("~/.rbenv/bin");

    // MacPorts
    try addPath("/opt/local/bin");
    try addPath("/opt/local/sbin");

    {
        var it = std.mem.split(u8, path, ":");
        while (it.next()) |p| {
            try addPath(p);
        }
    }

    // {
    //     for (paths.keys()) |p| {
    //         std.debug.print("PATH: {s}\n", .{p});
    //     }
    // }

    try stdout.print("{s}", .{output.items[0..(output.items.len - 1)]});
}

File: src/tools/pathHelper/pathHelper.go
package main

import (
	"os"
	"path"
	"strings"
)

var base = "/"
var output = ""
var seenPaths = make(map[string]struct{}, 64)

func addPath(path string) {
	if _, ok := seenPaths[path]; ok {
		return
	}

	seenPaths[path] = struct{}{}
	if strings.HasPrefix(path, "~/") {
		output += base
		path = path[2:]
	}

	output += path
	output += ":"
}

func main() {
	base = os.Args[1]
	envPath, ok := os.LookupEnv("PATH")
	if !ok {
		panic("missing PATH variable")
	}

	cfgDir, ok := os.LookupEnv("CFG_DIR")
	if !ok {
		panic("missing CFG_DIR variable")
	}

	{
		localPath := path.Join(cfgDir, "local", "path")
		addPath(localPath)
	}

	// linux gopath
	addPath("/usr/local/go/bin")

	// gopath
	addPath("~/go")
	addPath("~/go/bin")

	addPath("/opt/homebrew/bin")
	addPath("/opt/homebrew/Cellar/llvm/14.0.6/bin:$PATH")

	addPath("~/.rbenv/bin")

	// MacPorts
	addPath("/opt/local/bin")
	addPath("/opt/local/sbin")

	{
		for _, p := range strings.Split(envPath, ":") {
			addPath(p)
		}
	}

	print(output)
}
